contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py

```diff
@@ -0,0 +1,222 @@
+
+kept_family_accessions = []
+for i in range(1, 101):
+  kept_family_accessions.append('PF00' + '0'*(3-len(str(i))) + str(i))
+
+def keep_family_accessions(df, kept_family_accessions):
+  
+  dfs_by_family_accessions = []
+  
+  for family_accession in kept_family_accessions:
+    df_by_family_accession = df[df['mod_family_accession']==family_accession]
+    dfs_by_family_accessions.append(df_by_family_accession)
+  
+  return pd.concat(dfs_by_family_accessions)
+
+train_df = keep_family_accessions(train_df, kept_family_accessions)
+test_df = keep_family_accessions(test_df, kept_family_accessions)
+dev_df = keep_family_accessions(dev_df, kept_family_accessions)
+
+family_ids = sorted(set(train_df['family_id'].values))
+family_id_to_index = {}
+for i, family_id in enumerate(family_ids):
+  family_id_to_index[family_id] = i
+
+train_df['index'] = train_df.family_id.apply(lambda x: family_id_to_index[x])
+test_df['index'] = test_df.family_id.apply(lambda x: family_id_to_index[x])
+dev_df['index'] = dev_df.family_id.apply(lambda x: family_id_to_index[x])
+
+test_indexes = test_df['index'].values
+
+num_families = len(set(train_df['family_id']))
+
+def pad_seq(seq, pad_char='-'):
+  SEQ_LEN = 512
+  seq = seq[:SEQ_LEN]
+  padded_seq = seq + pad_char*(SEQ_LEN-len(seq))
+  return padded_seq
+
+AMINO_ACID_VOCABULARY = [
+    'A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R',
+    'S', 'T', 'V', 'W', 'Y', 'X', 'U', 'B', 'O', 'Z', '-'
+]
+num_categories = len(AMINO_ACID_VOCABULARY)
+
+def residues_to_one_hot_inds(amino_acid_residues):
+  return np.array([AMINO_ACID_VOCABULARY.index(char) for char in amino_acid_residues])
+
+train_df['one_hot_inds'] = train_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+test_df['one_hot_inds'] = test_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+dev_df['one_hot_inds'] = dev_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+
+print(train_df.head())
+
+
+"""Data batching."""
+def create_data_iterator(df, batch_size, epochs=1, buffer_size=None, seed=0, drop_remainder=False):
+
+  if buffer_size is None:
+    buffer_size = len(df)
+
+  one_hot_inds = list(df['one_hot_inds'].values)
+  one_hot_inds = tf.data.Dataset.from_tensor_slices(one_hot_inds)
+
+  indexes = df['index'].values
+  indexes = tf.data.Dataset.from_tensor_slices(indexes)
+
+  batches = tf.data.Dataset.zip((one_hot_inds, indexes)).shuffle(buffer_size=buffer_size, seed=seed)
+  batches = batches.repeat(epochs).batch(batch_size=batch_size, drop_remainder=drop_remainder).as_numpy_iterator()
+
+  return batches
+
+
+"""Model evaluation."""
+def evaluate(predict_fn, test_data, true_indexes, title, loss_fn_kwargs):
+
+  pred_indexes = []
+  cross_entropy = 0.
+  
+  for batch in iter(test_data):
+
+    X, Y = batch
+
+    Y_hat = predict_fn(X)
+
+    cross_entropy += cross_entropy_loss(Y, Y_hat, **loss_fn_kwargs)
+
+    preds = jnp.argmax(jax.nn.log_softmax(Y_hat), axis=1)
```

Comment:
the softmax is unnecessary here: it maintains the relative orders, so the argmax can take just Y_hat